NICE.Identity.Authorisation.WebAPI/Repositories/Context.cs

```diff
@@ -1,288 +1,322 @@
-			        .Where(ueh => EF.Functions.Like(ueh.EmailAddress, $"%{filter}%"))
-			        .Select(ueh => ueh.UserId.Value)
-			        .ToList();
-
-	        return Users.Where(u => (u.FirstName != null && EF.Functions.Like(u.FirstName, $"%{filter}%"))
-	                                || (u.LastName != null && EF.Functions.Like(u.LastName, $"%{filter}%"))
-	                                || (u.FirstName != null && u.LastName != null && EF.Functions.Like(u.FirstName + " " + u.LastName, $"%{filter}%"))
-	                                || (u.NameIdentifier != null && EF.Functions.Like(u.NameIdentifier, $"%{filter}%"))
-	                                || (u.EmailAddress != null && EF.Functions.Like(u.EmailAddress, $"%{filter}%"))
-	                                || (userIdsWithMatchingEmailHistory.Contains(u.UserId)))
-
-		        .Include(users => users.UserRoles)
-					.ThenInclude(userRoles => userRoles.Role)
-						.ThenInclude(website => website.Website)
-
-					.OrderByDescending(user => user.UserId)
-		        .ToList();
+
+        public List<User> GetUsers(IEnumerable<string> authenticationProviderUserIds)
+        {
+            var lowercasedAuthenticationProviderUserIds = authenticationProviderUserIds.Select(userId => userId.ToLower());
+	        return Users.Where(users => lowercasedAuthenticationProviderUserIds.Contains(users.NameIdentifier.ToLower()))
+		        .Include(users => users.UserRoles)
+		        .ThenInclude(userRoles => userRoles.Role)
+		        .ThenInclude(website => website.Website)
+		        .ToList();
+        }
+        internal IEnumerable<User> GetUsersByOrganisationId(int organisationId)
+        {
+            return Jobs.Where(jobs => jobs.OrganisationId.Equals(organisationId))
+                           .Include(users => users)
+                            .Select(x => x.User)
+                           .ToList();
+        }
+
+        internal UserAndJobForOrganisation GetUsersAndJobsByOrganisationId(int organisationId)
+        {
+            var jobs = Jobs.Where(jobs => jobs.OrganisationId.Equals(organisationId))
+                .Include(u => u.User)
+                .Include(o => o.Organisation)
+                .ToList();
+
+            
+            var usersAndJobs = new List<UserAndJobId>();
+            foreach (var job in jobs)
+            {
+                var userAndJobId = new UserAndJobId();
+                userAndJobId.UserId = job.UserId;
+                userAndJobId.User = new ApiModels.User(job.User);
+                userAndJobId.JobId = job.JobId;
+
+                usersAndJobs.Add(userAndJobId);
+            }
+
+            var usersForOrganisation = new UserAndJobForOrganisation();
+            usersForOrganisation.OrganisationId = organisationId;
+            usersForOrganisation.Organisation = new ApiModels.Organisation(jobs.FirstOrDefault().Organisation);
+            usersForOrganisation.Users = usersAndJobs;
+
+            return usersForOrganisation;
+        }
+
+

```

Comment:
I think it's possible to make this more concise by doing a projection, so something like this would give you everything rather than doing the separate bits i.e. getting jobs, then users and jobs and then users for organisation:

```
(from job in Jobs
                      where job.OrganisationId == organisationId
                      join user in Users on job.UserId equals user.UserId
                      join organisation in Organisations on job.OrganisationId equals organisation.OrganisationId
                      group job by new { job.JobId } into grp
                      select new UserAndJobForOrganisation
                      {
                          OrganisationId = organisationId,
                          Organisation = new ApiModels.Organisation(grp.Key.organisation),
                          Users = (List<UserAndJobId>)grp.Select(s => new UserAndJobId()
                          {
                              UserId = s.UserId,
                              User = new ApiModels.User(s.User),
                              JobId = s.JobId
                          })
                      }
                      ).ToList()
```

However, I've been trying to get it to work and it just won't work. I think it's because of a bug in EF 3.1 which hasn't been resolved:

https://github.com/dotnet/efcore/issues/20040

Anyway I just wanted to share it with you for awareness. I'll approve it now.

Reply:
:thumbsup: